### PR TITLE
Improve performance for #one? and #many?

### PR DESCRIPTION
**Pull request by gonzaloriestra**
`ActiveRecord::Relation` methods `one?` and `many?` generate a query like `SELECT COUNT(*) FROM posts` under the hood, and then compare if the result is equal (or greater) to 1. That count operation can be really slow for large tables or complex conditions, but there's no need to count all the records in these cases. It's much faster just by adding a limit, like `SELECT COUNT(count_column) FROM (SELECT 1 AS count_column FROM posts LIMIT 2)`. Other methods like `any?` do use a limit.

This is happening in all the Rails versions.

I've run some benchmarks and with 100M rows, the speed boost is more than 1300x.

<details>
  <summary>Benchmark code</summary>

```

# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "v6.1.3.1"
  gem "sqlite3"
  gem "benchmark-ips"
end

require "active_record"

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: "benchmark.sqlite3")

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.boolean :posted
  end
end

class Post < ActiveRecord::Base
  def self.fast_one?
    Post.limit(2).count(:all) == 1
  end
end

def benchmark
  puts
  puts "Benchmark with #{Post.count} rows"
  Benchmark.ips do |x|
    x.report("one?")        { Post.one? }
    x.report("fast_one?")   { Post.fast_one? }
    x.compare!
  end
end

print "Populating database"
names = [{ posted: true }] * 100_000
100.times do
  Post.insert_all(names)
  print "."
end

benchmark
Post.where('id > 10000000').delete_all
benchmark
Post.where('id > 1000000').delete_all
benchmark
Post.where('id > 100000').delete_all
benchmark


```

</details>

<details>
  <summary>Benchmark results</summary>

```
Benchmark with 100,000,000 rows
Warming up --------------------------------------
                one?     1.000  i/100ms
           fast_one?   518.000  i/100ms
Calculating -------------------------------------
                one?      3.832  (± 0.0%) i/s -     20.000  in   5.219749s
           fast_one?      5.084k (± 7.4%) i/s -     25.900k in   5.129227s

Comparison:
           fast_one?:     5084.1 i/s
                one?:        3.8 i/s - 1326.68x  (± 0.00) slower


Benchmark with 10,000,000 rows
Warming up --------------------------------------
                one?     3.000  i/100ms
           fast_one?   470.000  i/100ms
Calculating -------------------------------------
                one?     32.466  (±15.4%) i/s -    159.000  in   5.053688s
           fast_one?      4.466k (± 6.9%) i/s -     22.560k in   5.076130s

Comparison:
           fast_one?:     4466.3 i/s
                one?:       32.5 i/s - 137.57x  (± 0.00) slower


Benchmark with 1,000,000 rows
Warming up --------------------------------------
                one?    31.000  i/100ms
           fast_one?   402.000  i/100ms
Calculating -------------------------------------
                one?    375.534  (±13.3%) i/s -      1.860k in   5.067453s
           fast_one?      4.402k (±16.9%) i/s -     21.306k in   5.021693s

Comparison:
           fast_one?:     4401.5 i/s
                one?:      375.5 i/s - 11.72x  (± 0.00) slower


Benchmark with 100,000 rows
Warming up --------------------------------------
                one?   479.000  i/100ms
           fast_one?   447.000  i/100ms
Calculating -------------------------------------
                one?      5.479k (±14.0%) i/s -     27.303k in   5.106124s
           fast_one?      4.460k (± 6.2%) i/s -     22.350k in   5.033595s

Comparison:
                one?:     5479.4 i/s
           fast_one?:     4459.8 i/s - same-ish: difference falls within error

```

</details>

### CODE REVIEW

**Comment by byroot** [comment]
I can't put my finger on it, but I feel like I heard about this idea not so long ago.

Either way it sounds good to me. Could you please clean your commits? There's an extra merge commit that shouldn't be there.

**Comment by eugeneius** [comment]
https://github.com/rails/rails/pull/27175 was the previous attempt at this. Matthew's comment about the result no longer being memoised is still valid here.

**Comment by byroot** [comment]
Right, at the very least that "partial count" should be memoized as to make sure subsequence calls don't trigger a query.

**Comment by gonzaloriestra** [comment]
Thanks for the comments! I didn't know about that attempt, sorry. I had a look at previous issues/PRs, but there's too much content here... :sweat_smile: 

Regarding the memoization, the query is avoided when it's already loaded (see [this line](https://github.com/rails/rails/pull/42143/files#diff-18a561656864ea240daf46bdb1f50faace49f9ef74b90bcf667d9bbb17fce084R297)). I think it's covered by [this test](https://github.com/rails/rails/blob/70c5496542e5dc82ca28840cb01e710200ce5d14/activerecord/test/cases/associations/has_many_associations_test.rb#L2351). Or am I missing something?

**Comment by gonzaloriestra** [comment]
Sorry, I messed up a bit with the rebase and the bot added some wrong labels... Could you please remove them?

**Comment by byroot** [comment]
> Regarding the memoization, the query is avoided when it's already loaded (see [this line](https://github.com/rails/rails/pull/42143/files#diff-18a561656864ea240daf46bdb1f50faace49f9ef74b90bcf667d9bbb17fce084R297)). I think it's covered by [this test](https://github.com/rails/rails/blob/70c5496542e5dc82ca28840cb01e710200ce5d14/activerecord/test/cases/associations/has_many_associations_test.rb#L2351). Or am I missing something?

Yes, the memoization we are talking about is: `3.times { scope.one? }`, which would trigger 3 queries.

I think the result of `limit(2).count` could be memoized and re-used. It could even stored as the `count` memoization if it returns `1` or `0`. But maybe that's too brittle.

**Comment by gonzaloriestra** [comment]
But that was also triggering 3 queries before, wasn't it? And even with [size](https://github.com/rails/rails/blob/70c5496542e5dc82ca28840cb01e710200ce5d14/activerecord/lib/active_record/relation.rb#L265)... :thinking: 

Anyway, I agree. I've finally memoized the `limit(2).count` to share it between both methods. What do you think?

**Comment by byroot** [comment]
Hum, my bad. I assumed `size` was memoized as well, but it seems I was wrong. That's what Matthew said in https://github.com/rails/rails/pull/27175#issuecomment-431271531, did that change recently?

**Comment by gonzaloriestra** [comment]
I think it hasn't changed, it was similar in that moment: https://github.com/marshall-lee/rails/blob/046c3b6edf28d9604eb99b09bfb9501d5c50eee1/activerecord/lib/active_record/relation.rb#L240

**Comment by byroot** [comment]
Hum, then cc @matthewd, maybe you remember what you ment by "memoization" at the time?

**Comment by gonzaloriestra** [comment]
To ensure this PR doesn't fall into oblivion like the last attempt, what about merging the changes without memoization (as it was before), and opening a new issue to discuss adding it to all the convenient places in `Relation` (and not only to `one?` and `many?`)?

File: activerecord/lib/active_record/relation.rb
Comment (byroot) [inline activerecord/lib/active_record/relation.rb:309]: This should be private.

**Comment by byroot** [comment]
Ok, I looked at the code at the time of @matthewd's comment, and I'm pretty sure `#size` wasn't memoizing:

```
    # Returns size of the records.
    def size
      loaded? ? @records.length : count(:all)
    end
```

So I think this PR isn't degrading any performance in existing scenarios. So I'm 👍 . I'll merge once the method is made private and the merge conflict resolved.

**Review by byroot — DISMISSED** [review]

**Comment by byroot** [comment]
Wait, I'm having a doubt here, does this even work?  I'm more familiar with MySQL, so maybe it works with postgres, but `SELECT COUNT(*) LIMIT 2` will limit the number of returned rows, not the number of counted rows, right?

That's why the old PR is much more complicated, it does `SELECT 1 AS one LIMIT` and check the number of returned rows.

**Comment by gonzaloriestra** [comment]
Right, `SELECT COUNT(*) LIMIT 2` doesn't limit the scanned rows (at least in MySQL and PostgreSQL). I wrote that simplified query in the description, but it's not accurate (I've updated it).

The generated query is not that one. `Post.limit(2).count` is running the complex query in fact: `SELECT COUNT(count_column) FROM (SELECT 1 AS count_column FROM `posts` LIMIT 2) subquery_for_count`.

If you have a look at the benchmark, I think it's quite clear that it does work.

**Comment by byroot** [comment]
Ok, nevermind you are right. `limit(2).count` will generate a subquery: `SELECT COUNT(*) FROM (SELECT 1 AS one FROM "posts" LIMIT 2) subquery_for_count`. I just tested it.

My bad.

**Comment by matthewd** [comment]
Sorry I missed this earlier -- yeah, it seems I was confidently incorrect about `size` memoizing :confused:

**Pull request merged**
